framework/src/main/java/org/checkerframework/framework/type/DefaultTypeHierarchy.java

```diff
@@ -154,6 +154,25 @@ protected StructuralEqualityComparer createEqualityComparer() {
     @Override
     public boolean isSubtype(
             final AnnotatedTypeMirror subtype, final AnnotatedTypeMirror supertype) {
+        if (subtype.atypeFactory.ignoreUninferredTypeArguments
+                && (subtype.containsUninferredTypeArguments()
+                        || supertype.containsUninferredTypeArguments())) {
+            if (ignoreUninferredTypeArgument(subtype) || ignoreUninferredTypeArgument(supertype)) {
+                // If either is an uninferred type argument don't check primary annotations.
+                return true;
+            }
+            // Check primary annotations.
+            for (final AnnotationMirror top : qualifierHierarchy.getTopAnnotations()) {
+                AnnotationMirror subAnno = subtype.getAnnotationInHierarchy(top);
+                AnnotationMirror superAnno = supertype.getAnnotationInHierarchy(top);
+                if (subAnno != null
+                        && superAnno != null
+                        && !qualifierHierarchy.isSubtype(subAnno, superAnno)) {
+                    return false;
+                }
+            }
+            return true;
```

Comment:
I'm wondering whether this is a bit too coarse-grained.
With this change, even if somewhere deep within a generic type there is an uninferred type argument, only the primary annotations are checked.
Let's take a type `@A M<@B C, @D ?>` where the second type argument is uninferred.
With this change, only `@A` is checked. Wouldn't it be better if also `@B` is checked?
The first type argument isn't affected by the failure to infer the second argument.

What happens if you remove lines 157--159 and 164--174 from the patch? Methods like `isContainedBy` already check for uninferred type arguments, so maybe only the check in `isSubtype` was missing?
Then you wouldn't even need the new `containsUninferredTypeArguments` method.